File: src/main/java/com/pyamsoft/zaptorch/app/service/VolumeMonitorService.java
/*
 * Copyright 2016 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.zaptorch.app.service;

import android.accessibilityservice.AccessibilityService;
import android.content.Intent;
import android.os.Build;
import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.KeyEvent;
import android.view.accessibility.AccessibilityEvent;
import com.pyamsoft.zaptorch.ZapTorch;
import timber.log.Timber;

public class VolumeMonitorService extends AccessibilityService
    implements VolumeServicePresenter.VolumeServiceView {

  static VolumeMonitorService instance;
  private VolumeServicePresenter presenter;

  @CheckResult @NonNull static synchronized VolumeMonitorService getInstance() {
    if (instance == null) {
      throw new NullPointerException("VolumeMonitorService instance is NULL");
    }
    return instance;
  }

  static synchronized void setInstance(@Nullable VolumeMonitorService i) {
    instance = i;
  }

  public static void finish() {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
      getInstance().disableSelf();
    }
  }

  @CheckResult public static boolean isRunning() {
    return instance != null;
  }

  public static void changeCameraApi() {
    final VolumeMonitorService currentInstance = getInstance();
    // Simulate the lifecycle for destroying and re-creating the presenter
    Timber.d("Change camera API");
    currentInstance.getPresenter().unbindView();
    currentInstance.getPresenter().bindView(currentInstance);
  }

  @CheckResult @NonNull VolumeServicePresenter getPresenter() {
    return presenter;
  }

  @Override protected boolean onKeyEvent(KeyEvent event) {
    final int action = event.getAction();
    final int keyCode = event.getKeyCode();
    presenter.handleKeyEvent(action, keyCode);

    // Never consume events
    return false;
  }

  @Override public void onAccessibilityEvent(AccessibilityEvent accessibilityEvent) {
    Timber.d("onAccessibilityEvent");
  }

  @Override public void onInterrupt() {
    Timber.e("onInterrupt");
  }

  @Override protected void onServiceConnected() {
    super.onServiceConnected();

    if (presenter == null) {
      presenter = ZapTorch.get(this).provideComponent().provideVolumeServiceModule().getPresenter();
    }

    presenter.bindView(this);
    setInstance(this);
  }

  @Override public boolean onUnbind(Intent intent) {
    presenter.unbindView();
    setInstance(null);
    return super.onUnbind(intent);
  }

  @Override public void onDestroy() {
    super.onDestroy();
    presenter.destroy();
  }
}

File: app/src/main/java/com/pyamsoft/zaptorch/service/VolumeMonitorService.kt
/*
 * Copyright 2019 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.pyamsoft.zaptorch.service

import android.accessibilityservice.AccessibilityService
import android.content.Intent
import android.os.Build
import android.view.KeyEvent
import android.view.accessibility.AccessibilityEvent
import com.pyamsoft.pydroid.ui.Injector
import com.pyamsoft.zaptorch.ZapTorchComponent
import com.pyamsoft.zaptorch.api.CameraInterface.CameraError
import com.pyamsoft.zaptorch.service.ServiceControllerEvent.Finish
import com.pyamsoft.zaptorch.service.ServiceControllerEvent.RenderError
import com.pyamsoft.zaptorch.service.error.CameraErrorExplanation
import javax.inject.Inject
import timber.log.Timber

class VolumeMonitorService : AccessibilityService() {

    @JvmField
    @Inject
    internal var binder: ServiceBinder? = null

    override fun onKeyEvent(event: KeyEvent): Boolean {
        val action = event.action
        val keyCode = event.keyCode
        binder?.handleKeyEvent(action, keyCode)

        // Never consume events
        return false
    }

    override fun onAccessibilityEvent(accessibilityEvent: AccessibilityEvent) {
        Timber.d("onAccessibilityEvent")
    }

    override fun onInterrupt() {
        Timber.e("onInterrupt")
    }

    override fun onCreate() {
        super.onCreate()
        Injector.obtain<ZapTorchComponent>(applicationContext)
            .plusServiceComponent()
            .create()
            .inject(this)

        requireNotNull(binder).bind {
            return@bind when (it) {
                is RenderError -> renderError(it.error)
                is Finish -> finish()
            }
        }
    }

    private fun renderError(cameraError: CameraError) {
        applicationContext.also {
            val intent = cameraError.intent
            intent.setClass(it, CameraErrorExplanation::class.java)
            it.startActivity(intent)
        }
    }

    private fun finish() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            disableSelf()
        }
    }

    override fun onServiceConnected() {
        super.onServiceConnected()
        requireNotNull(binder).start()
    }

    override fun onUnbind(intent: Intent): Boolean {
        binder?.stop()
        return super.onUnbind(intent)
    }

    override fun onDestroy() {
        super.onDestroy()
        binder?.unbind()
        binder = null
    }
}
